refactor(orc8r): Replace moriyoshi routewrapper

MoritzThomasHuebner opened the pull request:
<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

Replaces "github.com/moriyoshi/routewrapper" package with "github.com/vishvananda/netlink".

I also did refactored some of the existing logic into separate functions. Unfortunately, the individual commits may be somewhat messy.

## Test Plan

`routewrapper` uses the `ip route` command to retrieve routing table information, so we can manually compare the new routing table with the changes to the one we get with `ip route` and also the one we used to get with `netlink`. 

I spun up some docker images and the vagrant machines to add some more routing table entries.

In the image below, I used the IntelliJ debugger to look in the routing table before (left) and after (right) I made the changes. The order has changed somewhat because of how the information is retrieved. However, the order in the routing table does not matter. I highlighted in red the routing table entries that are identical.

![Manual comparison](https://user-images.githubusercontent.com/8889531/179510261-ca3acf72-1f67-49c2-84db-f1f7d69acee0.png)

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

MoritzThomasHuebner committed:
refactor(agw): Remove routewrapper dependency

First attempt, needs refactoring

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Removed unused code.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Added length check to addresses

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Removed unused code

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Simplified linking

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Simplified linking

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Removed old routewrapper code

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Tidied up routing table resolving

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Simplified routing table resolving

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Get routes recursively

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Rationalised code a bit more

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Added some documentation

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Rationalised code

Made host route variable names more consistent and simplified some expressions.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
chore(agw): Removed routewrapper

Made host route variable names more consistent and simplified some expressions.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

sebathomas commented on orc8r/gateway/go/status/host_info.go:
```diff
@@ -78,10 +78,7 @@ func init() {
-	if err == nil {
-		hostRoutes, _ = w.Routes()
-	}
+	hostRoutes, _ = netlink.RouteList(nil, 0)
```
Maybe you can use `netlink.RouteList(nil, netlink.FAMILY_V4)` here to avoid ipv6 addresses instead of filtering later.

MoritzThomasHuebner replied:
Turns out this is the right idea. I made the change.

MoritzThomasHuebner committed:
refactor(orc8r): Update go.mod files

Updates the go.mod files outside the orc8r

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(orc8r): Remove IPv6 specific logic

Instead use `netlink.FAMILY_V4` to filter IPv6 out from the start.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

sebathomas commented on orc8r/gateway/go/status/gateway_status.go:
```diff
@@ -147,43 +148,68 @@ func GetNetworkInfo() *NetworkInfo {
+
+		if src == "" {
+			// No source IP is stored in hostRoute, we try to resolve this later
+			unlinkedHostRoutes = append(unlinkedHostRoutes, hostRoute)
+			continue
```
Not sure if I'm missing something but if we build the map `linkIndexToIP` at the beginning of the function, couldn't we resolve the source IP here and then avoid the recursion?

```suggestion
			src = linkIndexToIP[hostRoute.LinkIndex]
```

MoritzThomasHuebner replied:
This isn't quite as easy, we need to fully populate the `interfaceToIP` map first to build `linkIndexToIP`. 

Since we do not know in which order we find the link indices, we always will need to iterate twice over the host routes. The first time we iterate to find all where the source IP is given, and the second time to try to find matching link indices for the host routes that have no stated source IP.

I chose recursion since it seemed like the simplest way to avoid redundant logic.

MoritzThomasHuebner replied:
I removed the recursion. It now looks also a bit different, so maybe have another look.

sebathomas commented on orc8r/gateway/go/status/gateway_status.go:
```diff
@@ -249,3 +275,66 @@ func GetGatewayStatus() *GatewayStatus {
+}
+
+// getSourceIP resolves the source IP as a string given a host route.
+// Returns an empty string if no source IP is given.
```
This function is small enough that the comment is not needed. It's as easy to see what it does from reading the code as from reading the comment (and that's a good thing :)).

Same for the functions below.

sebathomas reviewed (commented):
Tested that it works on my machine. :+1:

Just added some ideas for refactoring.

MoritzThomasHuebner committed:
refactor(orc8r): Simplified routing table logic

Trimmed some unnecessary stuff from the new functions, mostly just return strings directly now. Removed some unused information in `getNetInterfaceID` logic. Trimmed some long comment strings. Extracted `linkRoutes` function

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

sebathomas commented on orc8r/gateway/go/status/gateway_status.go:
```diff
@@ -147,43 +148,73 @@ func GetNetworkInfo() *NetworkInfo {
+	routes, unlinkedHostRoutes, interfaceToIP := getRoutes(hostRoutes)
+	unlinkedHostRoutes = linkRoutes(unlinkedHostRoutes, interfaceToIP)
+	// Get the routing table entries for the newly linked host routes.
+	// recursiveResolving is false because we can no longer resolve any new
```
This comment needs to be updated too, still has a reference to recursive resolving.

MoritzThomasHuebner replied:
Done

sebathomas reviewed (approved):
Looking good, there's just a left-over comment from the old version that needs to be adapted or removed.
